domain/model/service/service.go

```diff
@@ -709,3 +729,56 @@ func (s WatchableService) WatchModel(ctx context.Context, modelUUID coremodel.UU
 		}),
 	)
 }
+
+// WatchModelCloudCredential returns a new NotifyWatcher watching for changes that
+// result in the cloud spec for a model changing. The changes watched for are:
+// - updates to model cloud.
+// - updates to model credential.
+// - changes to the credential set on a model.
+// The following errors can be expected:
+// - [modelerrors.NotFound] when the model is not found.
+func (s *WatchableService) WatchModelCloudCredential(ctx context.Context, modelUUID coremodel.UUID) (watcher.NotifyWatcher, error) {
+	if err := modelUUID.Validate(); err != nil {
+		return nil, errors.Errorf("invalid model UUID watching model cloud credential: %w", err)
+	}
+
+	// Get the model's cloud and credential UUID to use in the watcher filters.
+	cloudUUID, credentialUUID, err := s.st.GetModelCloudAndCredential(ctx, modelUUID)
+	if err != nil {
+		return nil, errors.Errorf("getting model cloud and credential: %w", err)
+	}
+
+	// The mapper is used to filter out any model events where the credential UUID has not changed.
+	mapper := func(ctx context.Context, txnRunner database.TxnRunner, events []changestream.ChangeEvent) ([]changestream.ChangeEvent, error) {
+		// Get the current model credential UUID.
+		_, newCredentialUUID, err := s.st.GetModelCloudAndCredential(ctx, modelUUID)
+		if err != nil {
+			return nil, errors.Capture(err)
+		}
+		var out []changestream.ChangeEvent
+		for _, e := range events {
+			if e.Namespace() != "model" || credentialUUID != newCredentialUUID {
```

Comment:
Do we have any plans to make these enums long term? Seems like magic strings in the new watchers could prove painful?

Reply:
I have the same thought each time we start to write a watcher. I can't see any consts that have been defined for the namespaces. Would be good to add them.